otherlibs/stdune-unstable/comparable.ml

```diff
@@ -1,5 +1,122 @@
 module Make (Key : Map.Key) = struct
   module Key = Key
   module Map = Map.Make (Key)
-  module Set = Set.Make (Key) (Map)
+
+  module Set = struct
+    type elt = Key.t
+
+    type t = unit Map.t
+
+    let empty = Map.empty
+
+    let is_empty = Map.is_empty
+
+    let singleton x = Map.singleton x ()
+
+    let cardinal = Map.cardinal
+
+    let equal a b = Map.equal a b ~equal:(fun () () -> true)
+
+    let compare a b = Map.compare a b ~compare:(fun () () -> Eq)
+
+    let to_list = Map.keys
+
+    let to_list_map t ~f =
+      (* optimize if we get a right fold *)
+      Map.to_list_map t ~f:(fun a () -> f a)
+
+    let mem t x = Map.mem t x
+
+    let add t x = Map.set t x ()
+
+    let remove t x = Map.remove t x
+
+    exception Not_a_subset
+
+    let is_subset t ~of_ =
+      match
+        Map.merge t of_ ~f:(fun _key a b ->
+            match (a, b) with
+            | Some (), None -> raise_notrace Not_a_subset
+            | _ -> None)
+      with
+      | (_ : t) -> true
+      | exception Not_a_subset -> false
+
+    let iter t ~f = Map.iteri t ~f:(fun k () -> f k)
+
+    let fold t ~init ~f = Map.foldi t ~init ~f:(fun k () acc -> f k acc)
+
+    let map t ~f = fold t ~init:empty ~f:(fun x acc -> add acc (f x))
+
+    let for_all t ~f = Map.for_alli t ~f:(fun k () -> f k)
+
+    let exists t ~f = Map.existsi t ~f:(fun k () -> f k)
+
+    let filter t ~f = Map.filteri t ~f:(fun k () -> f k)
+
+    let partition t ~f = Map.partitioni t ~f:(fun k () -> f k)
+
+    let min_elt t = Map.min_binding t |> Option.map ~f:fst
+
+    let max_elt t = Map.max_binding t |> Option.map ~f:fst
+
+    let choose t = Map.choose t |> Option.map ~f:fst
+
+    let split t x =
+      let a, x, b = Map.split t x in
+      (a, Option.is_some x, b)
+
+    let union a b = Map.union a b ~f:(fun _k () () -> Some ())
+
+    let diff a b =
```

Comment:
I recall doing a PR like this before and @aalekseyev called out that this implementation of diff wasn't efficient. Is it still the case?

Reply:
Eventually, it'd be nice to have our own implem for maps and sets. @bobot you should feel free to look at that.

BTW, for now I'm more interested in #5211 which is conservative. This current PR is more exploratory.